presto-spark/src/main/java/com/facebook/presto/spark/planner/SparkRddPlanner.java

```diff
@@ -0,0 +1,254 @@
+import static com.google.common.collect.Iterables.getOnlyElement;
+import static java.util.Objects.requireNonNull;
+
+public class SparkRddPlanner
+{
+    private final JsonCodec<SparkTaskDescriptor> sparkTaskRequestJsonCodec;
+
+    @Inject
+    public SparkRddPlanner(JsonCodec<SparkTaskDescriptor> sparkTaskRequestJsonCodec)
+    {
+        this.sparkTaskRequestJsonCodec = requireNonNull(sparkTaskRequestJsonCodec, "sparkTaskRequestJsonCodec is null");
+    }
+
+    public JavaPairRDD<Integer, byte[]> createSparkRdd(
+            JavaSparkContext sparkContext,
+            Session session,
+            PreparedPlan preparedPlan,
+            PrestoSparkTaskCompilerFactory taskCompilerFactory,
+            CollectionAccumulator<byte[]> taskStatsCollector)
+    {
+        RddFactory rddFactory = new RddFactory(
+                session,
+                sparkTaskRequestJsonCodec,
+                sparkContext,
+                taskCompilerFactory,
+                getHashPartitionCount(session),
+                taskStatsCollector,
+                preparedPlan.getTableWriteInfo());
+        return rddFactory.createRdd(preparedPlan.getPlan());
+    }
+
+    private static class RddFactory
+    {
+        private final Session session;
+        private final JsonCodec<SparkTaskDescriptor> sparkTaskRequestJsonCodec;
+        private final JavaSparkContext sparkContext;
+        private final PrestoSparkTaskCompilerFactory compilerFactory;
+        private final int hashPartitionCount;
+        private final CollectionAccumulator<byte[]> taskStatsCollector;
+        private final TableWriteInfo tableWriteInfo;
+
+        private RddFactory(
+                Session session,
+                JsonCodec<SparkTaskDescriptor> sparkTaskRequestJsonCodec,
+                JavaSparkContext sparkContext,
+                PrestoSparkTaskCompilerFactory taskCompilerFactory,
+                int hashPartitionCount,
+                CollectionAccumulator<byte[]> taskStatsCollector,
+                TableWriteInfo tableWriteInfo)
+        {
+            this.session = requireNonNull(session, "session is null");
+            this.sparkTaskRequestJsonCodec = requireNonNull(sparkTaskRequestJsonCodec, "sparkTaskRequestJsonCodec is null");
+            this.sparkContext = requireNonNull(sparkContext, "sparkContext is null");
+            this.compilerFactory = requireNonNull(taskCompilerFactory, "taskCompilerFactory is null");
+            this.hashPartitionCount = hashPartitionCount;
+            this.taskStatsCollector = requireNonNull(taskStatsCollector, "taskStatsCollector is null");
+            this.tableWriteInfo = requireNonNull(tableWriteInfo, "tableWriteInfo is null");
+        }
+
+        public JavaPairRDD<Integer, byte[]> createRdd(SubPlanWithTaskSources subPlan)
+        {
+            PlanFragment fragment;
+            if (subPlan.getFragment().getPartitioningScheme().getPartitioning().getHandle().equals(FIXED_HASH_DISTRIBUTION)) {
```

Comment:
This should be done before creating RDD -- it should be consider as the `plan adaption` phase.  We can put a TODO to refactor later though.